compare_platforms/compare_platforms_read_only.c

```diff
@@ -0,0 +1,147 @@
+    printf("privatereview=%#p publicreview=%#p diff=%tx\n", privatereview, publicreview, publicreview - privatereview);
+    printf("publicreview=%#p permissions=%#p diff=%tx\n", publicreview, permissions, permissions - publicreview);
+    printf("realname=%#p permissions=%#p diff=%tx\n", realname, permissions, permissions - realname);
+#else
+    printf("\nusername=%#p realname=%#p diff=%tx\n", username, realname, realname - username);
+    printf("realname=%#p privatereview=%#p diff=%tx\n", realname, privatereview, privatereview - realname);
+    printf("privatereview=%#p publicreview=%#p diff=%tx\n", privatereview, publicreview, publicreview - privatereview);
+    printf("publicreview=%#p permissions=%#p diff=%tx\n", publicreview, permissions, permissions - publicreview);
+    printf("realname=%#p permissions=%#p diff=%tx\n", realname, permissions, permissions - realname);
+#endif
+
+    assert((size_t)(username + smallsz) <= (size_t)realname); 
+    assert((size_t)(privatereview + biggersz) <= (size_t)publicreview);
+    assert((size_t)(realname + smallsz) <= (size_t)permissions); 
+
+// If we are in a CHERI environment, protect the review by setting
+// the relevant struct members (in this case all of them)
+// to read-only.  
+bool b_improved = false;
+  
+#ifdef __CHERI_PURE_CAPABILITY__
+    if(can_write(&review) == false) {
+        printf("\nSetting the review to read-only.\n");
+        struct review *ro_review = set_read_only(&review);
+        assert((cheri_perms_get(ro_review) & CHERI_PERM_STORE) == 0);
+
+        printf("\nThe struct is read-only so trying to change the review will make the program crash...\n");
+        fflush(stdout);
+        
+        b_improved = try_improve_publicreview(ro_review, biggersz, false);
+        
+        // This line should be unreachable.
+        print_details(ro_review);
+    }
+# else
+    // Try editing the public review in a non-CHERI environment,
+    // by simply checking the write permissions and refusing if none are found.
+    b_improved = try_improve_publicreview(&review, biggersz, false);
+#endif
+
+    // This code will be unreachable in a CHERI environment,
+    // but it would crash whether or not the struct had been set to read-only.
+    if(b_improved == false) {
+        printf("\nOverflowing reviewer realname by 1\n");
+        memset(review.realname + smallsz+1, 'w', 1);
+    
+        // Now we can see the changes, if any. The permissions should have changed.
+        print_details(&review);
+    
+        // Overflow the realname even more, which in a CHERI
+        // environment should crash if it hasn't already.
+        const size_t oversz = review.permissions - review.realname + 2 - smallsz;
+        printf("\nOverflowing reviewer realname by %zx\n", oversz);
+        memset(review.realname + smallsz+2, 'w', oversz);
+    
+        print_details(&review);
+        
+        b_improved = try_improve_publicreview(&review, biggersz, false);
```

Comment:
Is there any difference between this and the call above?

Reply:
try_improve_publicreview has been removed and its equivalent functionality folded into change_publicreview, lines 50-61.

Reply:
> Is there any difference between this and the call above?

Fixed in a41b3e55231a43e96a0da7612f43668dd452f796